### PR TITLE
Sync node create

### PR DESCRIPTION
**Pull request by zafields**

### CODE REVIEW

**Comment by zafields** [comment]
Addresses issue #96 by making the node binding create synchronous, and updates the `build_nodejs.cmd` to properly create the `NODE_INCLUDE` and `NODE_LIB` environment variables.

**Comment by damonbarry** [comment]
@zafields there are build errors on [windows](http://azure-iot-sdks-ci.westus.cloudapp.azure.com:8080/job/new-win-c/2221//console).

File: bindings/nodejs/inc/nodejs_common.h [hunk 1]
```diff
@@ -55,6 +56,7 @@ struct NODEJS_MODULE_HANDLE_DATA
     NODEJS_MODULE_HANDLE_DATA(NODEJS_MODULE_HANDLE_DATA&& rhs)
     {
         broker = rhs.broker;
+        create_complete = std::move(rhs.create_complete);
```
Comment (avranju) [inline bindings/nodejs/inc/nodejs_common.h]: I am not sure, but did we not decide not to do this in the move constructor?

File: bindings/nodejs/src/nodejs.cpp [hunk 2]
```diff
@@ -110,7 +112,31 @@ static MODULE_HANDLE NODEJS_Create(BROKER_HANDLE broker, const void* configurati
                 }
                 else
                 {
-                    handle_data->SetModuleState(NodeModuleState::initializing);
+                    // Wait for the module to become fully initialized
+                    NodeModuleState module_state;
+                    std::future<NodeModuleState> create_gate = handle_data->create_complete.get_future();
```
Comment (avranju) [inline bindings/nodejs/src/nodejs.cpp]: Use `auto` instead of spelling out the type?

File: bindings/nodejs/src/nodejs.cpp [hunk 3]
```diff
@@ -110,7 +112,31 @@ static MODULE_HANDLE NODEJS_Create(BROKER_HANDLE broker, const void* configurati
...
+                    switch (create_gate.wait_for(std::chrono::seconds(10)))
```
Comment (avranju) [inline bindings/nodejs/src/nodejs.cpp]: Define a constant for the timeout period of `10` seconds? In fact the constant could perhaps be a `std::chrono` instance.

File: bindings/nodejs/src/nodejs.cpp [hunk 4]
```diff
@@ -110,7 +112,31 @@ static MODULE_HANDLE NODEJS_Create(BROKER_HANDLE broker, const void* configurati
...
+                    {
+                      case std::future_status::ready:
+                          module_state = create_gate.get();
+                        break;
```
Comment (avranju) [inline bindings/nodejs/src/nodejs.cpp]: Nit: `break` indentation should align with line 121. Here and on line 127.

File: bindings/nodejs/inc/nodejs_common.h [hunk 5]
```diff
@@ -140,16 +142,17 @@ struct NODEJS_MODULE_HANDLE_DATA
...
+    BROKER_HANDLE broker;
+    std::promise<NodeModuleState> create_complete;
+    std::string main_path;
+    std::string configuration_json;
+    v8::Isolate *v8_isolate;
+    v8::Persistent<v8::Object> module_object;
+    size_t module_id;
+    PFNMODULE_START on_module_start;
+    NodeModuleState module_state;
+    bool start_pending;
```
Comment (avranju) [inline bindings/nodejs/inc/nodejs_common.h]: The idea behind having this flag `start_pending` was to handle the case where a call to `NODEJS_Start` occurs before the module's state has transitioned to `initialized` from `initializing`. Now that `NODEJS_Create` is a blocking call and we no longer have an `initializing` state, we may not need this flag anymore. Consider removing this field and the associated logic.

File: tools/build_nodejs.cmd [hunk 6]
```diff
@@ -47,8 +47,7 @@ copy node\build\Release\lib\*.lib dist\lib
 
 popd
 
-rem Export environment variables for where the include/lib files can be found
-set "NODE_INCLUDE=%build-root%\dist\inc"
-set "NODE_LIB=%build-root%\dist\lib"
+rem Set environment variables for where the include/lib files can be found
+@endlocal & set NODE_INCLUDE=%build-root%\dist\inc\ & set NODE_LIB=%build-root%\dist\lib\
```
Comment (avranju) [inline tools/build_nodejs.cmd]: Nice!

File: bindings/nodejs/src/nodejs.cpp (hunk 4 above)
Comment (zafields) [inline bindings/nodejs/src/nodejs.cpp]: fixed (overly helpful VS caused this).

File: bindings/nodejs/src/nodejs.cpp (hunk 2 above)
Comment (zafields) [inline bindings/nodejs/src/nodejs.cpp]: done.

File: bindings/nodejs/src/nodejs.cpp (hunk 3 above)
Comment (zafields) [inline bindings/nodejs/src/nodejs.cpp]: Just gave the number a name for now.

File: tools/build_nodejs.cmd (hunk 6 above)
Comment (zafields) [inline tools/build_nodejs.cmd]: 👍

File: bindings/nodejs/inc/nodejs_common.h (hunk 1 above)
Comment (zafields) [inline bindings/nodejs/inc/nodejs_common.h]: Regardless of the life of any individual data structure used to contain the handle data, the collection of handle data represents a single object. So I believe the future (along with all other data) should be moved if possible to reflect the intent of what is happening.
Unfortunately, we are only able to preserve this data in the MOVE constructor and not in the copy constructors. I see how this is more confusing than not bring them along all together, but I must admit I don't like the way it feels.

File: bindings/nodejs/inc/nodejs_common.h (hunk 5 above)
Comment (zafields) [inline bindings/nodejs/inc/nodejs_common.h]: removed with test passing

File: bindings/nodejs/inc/nodejs_common.h (hunk 1 above)
Comment (zafields) [inline bindings/nodejs/inc/nodejs_common.h]: I'm going to leave a comment to this effect and forget about it. :D

**Pull request merged**